### PR TITLE
Add code to construct pipelined loop from schedule

### PR DESCRIPTION
**Pull request by harsh-nod**
This PR adds code to construct the epilogue, kernel
and prologue once we have computed a schedule. We
simulate rotating registers in software and add
visualization tools to show the pipelined graphs.

### CODE REVIEW

File: shark_turbine/kernel/ops/wave_ops.py [hunk 1]
```diff
@@ -453,6 +453,8 @@ def index(self, value: Any):
             self.fx_node.index = {}
             for dim, key in value.items():
                 self.fx_node.index[dim] = key
+        elif isinstance(value, list):
+            self.fx_node.index = list(value)
```
Comment (Hardcode84) [inline shark_turbine/kernel/ops/wave_ops.py:457]: nit: you don't need `list(...)` if you already did `isinstance(value, list)`

File: lit_tests/kernel/wave/codegen.py [hunk 2]
```diff
@@ -613,6 +613,339 @@ def repeat(acc: tkl.Register[M, N, tkl.f32]) -> tkl.Register[M, N, tkl.f32]:
...
+        },
+        canonicalize=True,
+        schedule=True,
+    ):
+        a = torch.randn(64, 32, dtype=torch.float16)
+        b = torch.randn(128, 32, dtype=torch.float16)
+        c = torch.zeros(64, 128, dtype=torch.float32)
+        print(gemm_pipelined(a, b, c).module_op)
+
+        # CHECK:          func.func @gemm_pipelined(%[[ARG0:[a-zA-Z0-9_]+]]: !stream.binding, %[[ARG1:[a-zA-Z0-9_]+]]:
```
[the remark this answers is not in the document]
Comment (harsh-nod) [reply lit_tests/kernel/wave/codegen.py:678]: Sure, will just check for stores, loads and mma + for.

File: shark_turbine/kernel/wave/codegen.py [hunk 3]
```diff
@@ -921,7 +922,10 @@ def handle_reduction(emitter: WaveEmitter, node: fx.Node):
 
     # For now, we assume that dimensions that have tiling constraints on them,
     # do not have any other constraints.
-    end = arith_d.constant(IndexType.get(), int(count))
+    end_value = int(count)
```
[the remark this answers is not in the document]
Comment (harsh-nod) [reply shark_turbine/kernel/wave/codegen.py:925]: Sure let me look into this.

[the remark this answers is not in the document]
Comment (harsh-nod) [reply shark_turbine/kernel/wave/codegen.py]: okay I have changed this.

**Review by Hardcode84 — APPROVED** [review]

**Pull request merged**